File: automation-agent/platform-fastapi-server/app/utils/case_collection.py
"""
测试用例集合工具
从 Flask 迁移到 FastAPI
"""
import os
import json
from datetime import datetime
from typing import Dict, Any, List


class CaseCollection:
    """测试用例集合工具类"""
    
    @staticmethod
    def generate_history_info(
        coll_type: str,
        case_collection_info: Dict[str, Any],
        report_data_path: str,
        execute_uuid: str
    ) -> List[Dict[str, Any]]:
        """
        生成历史记录信息
        
        Args:
            coll_type: 用例类型
            case_collection_info: 用例集合信息
            report_data_path: 报告数据路径
            execute_uuid: 执行 UUID
        
        Returns:
            List[Dict]: 历史记录列表
        """
        # 获取报告数据路径的所有文件列表
        file_list = os.listdir(report_data_path)
        history_list = []
        
        for file in file_list:
            # 确定是 result.json 文件
            if file.endswith("result.json"):
                with open(os.path.join(report_data_path, file), 'r', encoding='utf-8') as f:
                    data = json.load(f)
                    
                    if data:
                        # 获取第一层数据
                        name = data.get("name", "")
                        status = data.get("status", "")
                        start = data.get("start", 0)
                        stop = data.get("stop", 0)
                        
                        # 计算执行时长（毫秒）
                        duration = stop - start
                        
                        # 构建历史记录字典
                        history_record = {
                            "name": name,
                            "status": status,
                            "type": coll_type,  # 使用常量代替硬编码
                            "project_id": case_collection_info.get("project_id"),
                            "coll_id": case_collection_info.get("id"),
                            "duration": duration,
                            # 添加创建时间，为当前时间
                            "create_time": datetime.strftime(datetime.today(), '%Y-%m-%d %H:%M:%S'),
                            # 添加 detail 字段为 execute_uuid
                            "detail": execute_uuid
                        }
                        
                        history_list.append(history_record)
        
        return history_list
    
    @staticmethod
    def parse_result_file(result_file_path: str) -> Dict[str, Any]:
        """
        解析 result.json 文件
        
        Args:
            result_file_path: result.json 文件路径
        
        Returns:
            Dict: 解析后的数据
        """
        try:
            with open(result_file_path, 'r', encoding='utf-8') as f:
                data = json.load(f)
                return data
        except Exception as e:
            print(f"解析结果文件失败: {e}")
            return {}
    
    @staticmethod
    def list_report_files(report_dir: str) -> List[str]:
        """
        列出报告目录中的所有文件
        
        Args:
            report_dir: 报告目录路径
        
        Returns:
            List[str]: 文件列表
        """
        try:
            return os.listdir(report_dir)
        except Exception as e:
            print(f"列出报告文件失败: {e}")
            return []
    
    @staticmethod
    def create_history_records(
        coll_type: str,
        case_collection_info: Dict[str, Any],
        execution_data: Dict[str, Any]
        report_dir: str
    ) -> Dict[str, Any]:
        """
        创建历史记录
        
        Args:
            coll_type: 用例类型
            case_collection_info: 用例集合信息
            execution_data: 执行数据
            report_dir: 报告目录
        
        Returns:
            Dict: 创建的历史记录
        """
        return {
            "name": case_collection_info.get("collection_name", ""),
            "status": execution_data.get("status", "completed"),
            "type": coll_type,
            "project_id": case_collection_info.get("project_id"),
            "coll_id": case_collection_info.get("id"),
            "duration": execution_data.get("duration", 0),
            "create_time": datetime.strftime(datetime.today(), '%Y-%m-%d %H:%M:%S'),
            "detail": execution_data.get("execute_uuid", "")
        }

File: automation-agent/platform-java-server/src/main/java/com/platform/platformserver/utils/CaseCollection.java
package com.platform.platformserver.utils;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Component;

import java.io.File;
import java.io.IOException;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.*;

@Slf4j
@Component
public class CaseCollection {

    private final ObjectMapper objectMapper = new ObjectMapper();
    private static final DateTimeFormatter formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss");

    public List<Map<String, Object>> generateHistoryInfo(String collType, Map<String, Object> caseCollectionInfo, 
                                                           String reportDataPath, String executeUuid) {
        List<Map<String, Object>> historyList = new ArrayList<>();
        
        try {
            File reportDir = new File(reportDataPath);
            if (!reportDir.exists() || !reportDir.isDirectory()) {
                return historyList;
            }

            File[] files = reportDir.listFiles((dir, name) -> name.endsWith("result.json"));
            if (files == null) {
                return historyList;
            }

            for (File file : files) {
                try {
                    JsonNode data = objectMapper.readTree(file);
                    if (data != null && !data.isEmpty()) {
                        String name = data.has("name") ? data.get("name").asText() : "";
                        String status = data.has("status") ? data.get("status").asText() : "";
                        long start = data.has("start") ? data.get("start").asLong() : 0;
                        long stop = data.has("stop") ? data.get("stop").asLong() : 0;
                        long duration = stop - start;

                        Map<String, Object> historyRecord = new HashMap<>();
                        historyRecord.put("name", name);
                        historyRecord.put("status", status);
                        historyRecord.put("type", collType);
                        historyRecord.put("project_id", caseCollectionInfo.get("project_id"));
                        historyRecord.put("coll_id", caseCollectionInfo.get("id"));
                        historyRecord.put("duration", duration);
                        historyRecord.put("create_time", LocalDateTime.now().format(formatter));
                        historyRecord.put("detail", executeUuid);

                        historyList.add(historyRecord);
                    }
                } catch (IOException e) {
                    log.error("解析结果文件失败: {}", file.getName(), e);
                }
            }
        } catch (Exception e) {
            log.error("生成历史记录信息失败", e);
        }

        return historyList;
    }

    public Map<String, Object> parseResultFile(String resultFilePath) {
        try {
            File file = new File(resultFilePath);
            if (file.exists()) {
                JsonNode data = objectMapper.readTree(file);
                @SuppressWarnings("unchecked")
                Map<String, Object> result = (Map<String, Object>) objectMapper.convertValue(data, Map.class);
                return result;
            }
        } catch (Exception e) {
            log.error("解析结果文件失败: {}", resultFilePath, e);
        }
        return new HashMap<>();
    }

    public List<String> listReportFiles(String reportDir) {
        try {
            File dir = new File(reportDir);
            if (dir.exists() && dir.isDirectory()) {
                File[] files = dir.listFiles();
                if (files != null) {
                    List<String> fileList = new ArrayList<>();
                    for (File file : files) {
                        fileList.add(file.getName());
                    }
                    return fileList;
                }
            }
        } catch (Exception e) {
            log.error("列出报告文件失败: {}", reportDir, e);
        }
        return new ArrayList<>();
    }

    public Map<String, Object> createHistoryRecords(String collType, Map<String, Object> caseCollectionInfo,
                                                   Map<String, Object> executionData, String reportDir) {
        Map<String, Object> historyRecord = new HashMap<>();
        historyRecord.put("name", caseCollectionInfo.get("collection_name"));
        historyRecord.put("status", executionData.getOrDefault("status", "completed"));
        historyRecord.put("type", collType);
        historyRecord.put("project_id", caseCollectionInfo.get("project_id"));
        historyRecord.put("coll_id", caseCollectionInfo.get("id"));
        historyRecord.put("duration", executionData.getOrDefault("duration", 0));
        historyRecord.put("create_time", LocalDateTime.now().format(formatter));
        historyRecord.put("detail", executionData.get("execute_uuid"));
        return historyRecord;
    }
}
